### ISSUE TITLE
show(::QRCompactWYQ) doesn't work properly

### ISSUE THREAD

**Issue by jiahao** [JuliaLang/LinearAlgebra.jl#127]
``` julia
julia> qrfact(eye(2))[:Q] |> display
2x2 QRCompactWYQ{Float64}:
 1.0  0.0
 0.0  1.0

julia> qrfact(eye(2))[:Q] |> show
[#undef #undef
 #undef #undef]
```

cc: @andreasnoackjensen - is this because `getindex(::QRCompactWYQ, i,j)` isn't defined?

**Comment by andreasnoack** [comment]
At least, it is sufficient to define a `getindex` method to get `show` working. Maybe we should define a general `getindex{AbstractMatrix{<:Number}}` method that works by pre and post multiplying unit vectors.

**Comment by jiahao** [comment]
I like the idea of a general `getindex`.

**Comment by andreasnoack** [comment]
I'll do that. Actually, I have done it, but I need to investigate some ambiguity warnings to figure out which `could not show value of type Tuple` Julia is thinking of.

**Comment by JeffBezanson** [comment]
I think a type that can't implement `getindex` is not really an array. I'm not sure we should suddenly introduce the idea that for all conceivable arrays, multiplying by vectors is more fundamental than getting elements.

**Comment by andreasnoack** [comment]
even when restricted to `AbstractMatrix{<:Number}`?

**Comment by jiahao** [comment]
From the offline discussion, the idea was that `QRCompactWYQ` isn't actually an `AbstractArray` since it isn't a container that can implement `getindex` in less than `O(N)` time (ref JuliaLang/julia#987).

I guess the real issue is why `show()` is defined the way it is; maybe it should imitate `display()` and dump the raw data.

**Comment by andreasnoack** [comment]
This was the reason to have `StoredArray<:AbstractArray`. See https://github.com/JuliaLang/julia/issues/6212. If the special storage matrices cannot be allowed to be abstract matrices anymore I think we'd need some kind of `AbstractParametrized{T}` type to make it possible to define something like the `Transpose` type in JuliaLang/julia#6837.

**Comment by andreasnoack** [comment]
Fixed by 8e76a21b5a62266eee445cfe9c9e12b6a0600509
